### ISSUE TITLE
Providers to support Promises/Observables

### ISSUE THREAD

**Issue by raygig** [abacritt/angularx-social-login#122]
Our application gets all the app configurations from the backend via an API call. Would it be possible for your library to detect a promise and wait for the resolution of that promise to populate the provider config values? If it can also support Observables that would be great as well.

For example here is what the provider configuration could look like.

```
export function provideConfig(configService) {

  return new Promise((resolve) => {

    configService.get()
    .subscribe((response) => {

      const config = new AuthServiceConfig([
        {
          id: GoogleLoginProvider.PROVIDER_ID,
          provider: new GoogleLoginProvider(response.googleClientId)
        },
        {
          id: FacebookLoginProvider.PROVIDER_ID,
          provider: new FacebookLoginProvider(response.facebookClientId)
        }
      ]);

      resolve(config);
    });
  });
}

```

```
@NgModule({
  providers: [
    { provide: AuthServiceConfig, useFactory: provideConfig }
  ]
})
```

**Comment by ips-illuminz** [comment]
I have also the same requirement, if i have to change api keys from backend then i need to change in frontend and make new build. It will be helpful to set settings with promise or settings from api call.

**Comment by jaibatrik** [comment]
Would be great if somebody can send a PR!

**Comment by somcoder** [comment]
Have you guys looked at Angular's APP_INITIALIZER token?
I had this exact same scenario, and here is how I approached it:

``` typescript
// localStorage key.
const authConfigStoreKey = 'auth-config';

// Your configuration provider.
export function provideConfig() {
   // Get the auth provider configuration from localStorage.
    const authConfig = JSON.parse(localStorage.getItem(authConfigStoreKey) || '{}');
    
    return new AuthServiceConfig([
        {
            id: GoogleLoginProvider.PROVIDER_ID,
            provider: new GoogleLoginProvider(authConfig['google']['clientId'], { redirect_uri: 'http://localhost:4200' })
        },
        {
            id: FacebookLoginProvider.PROVIDER_ID,
            provider: new FacebookLoginProvider(authConfig['facebook']['appId'])
        }
    ]);
}

// App initializer. This is called whenever the app is initialized, and that's when we need to get the configurations from the server.
export function initApp(injector: Injector) {
    return () => {
        return new Promise<Boolean>(async (resolve, reject) => {
            // Load auth providers configuration.
            const authConfig = localStorage.getItem(authConfigStoreKey);
            if (!authConfig) {
                const result = await http.get('http://localhost:3000/auth-config').toPromise();
                if (result) {
                    localStorage.setItem(authConfigStoreKey, JSON.stringify(result));
                }
            }

            resolve(true);
        }
    }
}
```
And then add this to your `providers`
``` typescript
providers: [
    {
        provide: APP_INITIALIZER,
        useFactory: initApp,
        deps: [Injector],
        multi: true
    },
...
]
```
This would do it, happy coding. ☺️

**Comment by BrianWClausen** [comment]
This is a great solution, but is there someway of making this work without the use of localstorage? I need it to work with universal on a node server.

**Comment by somcoder** [comment]
@BrianWClausen I used `localStorage` as a caching mechanism, and I think you can use it too. [See this](https://stackoverflow.com/a/39098748/4672277)

**Comment by BrianWClausen** [comment]
@somcoder Thanks! I'll take a look.

**Comment by BrianWClausen** [comment]
The solution proposed earlier in the thread didn't work for me. I have created a pull request #247 to resolve the issue.

**Comment by jaibatrik** [comment]
Closed via #122.

**Comment by sterichards** [comment]
Could an example be added please?

I don't see any changes to the readme that describe implementing this change
